File: lib/features/life_stage/domain/life_stage.dart
enum LifeStageMode {
  cycleTracking,
  tryingToConceive,
  pregnancy,
  postpartum,
  breastfeeding,
  perimenopause,
  menopauseTransition,
  hormonalContraception,
}

extension LifeStageModeLabel on LifeStageMode {
  String get label => switch (this) {
    LifeStageMode.cycleTracking => 'Cycle tracking',
    LifeStageMode.tryingToConceive => 'Trying to conceive',
    LifeStageMode.pregnancy => 'Pregnancy',
    LifeStageMode.postpartum => 'Postpartum',
    LifeStageMode.breastfeeding => 'Breastfeeding',
    LifeStageMode.perimenopause => 'Perimenopause',
    LifeStageMode.menopauseTransition => 'Menopause transition',
    LifeStageMode.hormonalContraception => 'Hormonal contraception',
  };
}

class LifeStageCapabilities {
  const LifeStageCapabilities({
    required this.predictNextPeriod,
    required this.showFertilityObservations,
    required this.showPregnancyLogging,
    required this.preserveHistoricalCycles,
    required this.claimsContraceptiveEffectiveness,
  });

  final bool predictNextPeriod;
  final bool showFertilityObservations;
  final bool showPregnancyLogging;
  final bool preserveHistoricalCycles;
  final bool claimsContraceptiveEffectiveness;

  factory LifeStageCapabilities.forMode(LifeStageMode mode) {
    return switch (mode) {
      LifeStageMode.pregnancy => const LifeStageCapabilities(
        predictNextPeriod: false,
        showFertilityObservations: false,
        showPregnancyLogging: true,
        preserveHistoricalCycles: true,
        claimsContraceptiveEffectiveness: false,
      ),
      LifeStageMode.tryingToConceive => const LifeStageCapabilities(
        predictNextPeriod: true,
        showFertilityObservations: true,
        showPregnancyLogging: false,
        preserveHistoricalCycles: true,
        claimsContraceptiveEffectiveness: false,
      ),
      LifeStageMode.postpartum ||
      LifeStageMode.breastfeeding => const LifeStageCapabilities(
        predictNextPeriod: false,
        showFertilityObservations: false,
        showPregnancyLogging: false,
        preserveHistoricalCycles: true,
        claimsContraceptiveEffectiveness: false,
      ),
      LifeStageMode.menopauseTransition => const LifeStageCapabilities(
        predictNextPeriod: false,
        showFertilityObservations: false,
        showPregnancyLogging: false,
        preserveHistoricalCycles: true,
        claimsContraceptiveEffectiveness: false,
      ),
      LifeStageMode.cycleTracking ||
      LifeStageMode.perimenopause ||
      LifeStageMode.hormonalContraception => const LifeStageCapabilities(
        predictNextPeriod: true,
        showFertilityObservations: false,
        showPregnancyLogging: false,
        preserveHistoricalCycles: true,
        claimsContraceptiveEffectiveness: false,
      ),
    };
  }
}

File: web/src/features/life-stage/life-stage.ts
export const LIFE_STAGE_MODES = [
  'cycleTracking',
  'tryingToConceive',
  'pregnancy',
  'postpartum',
  'breastfeeding',
  'perimenopause',
  'menopauseTransition',
  'hormonalContraception',
] as const;

export type LifeStageMode = (typeof LIFE_STAGE_MODES)[number];

export const LIFE_STAGE_CAPABILITY_IDS = [
  'LIFE-001',
  'LIFE-002',
  'LIFE-003',
  'LIFE-004',
  'LIFE-005',
  'LIFE-006',
  'LIFE-007',
  'LIFE-008',
  'LIFE-009',
] as const;

export type LifeStageCapabilities = {
  predictNextPeriod: boolean;
  showFertilityObservations: boolean;
  showPregnancyLogging: boolean;
  preserveHistoricalCycles: true;
  claimsContraceptiveEffectiveness: false;
};

const LABELS: Record<LifeStageMode, string> = {
  cycleTracking: 'Cycle tracking',
  tryingToConceive: 'Trying to conceive',
  pregnancy: 'Pregnancy',
  postpartum: 'Postpartum',
  breastfeeding: 'Breastfeeding',
  perimenopause: 'Perimenopause',
  menopauseTransition: 'Menopause transition',
  hormonalContraception: 'Hormonal contraception',
};

const BASE: Pick<LifeStageCapabilities, 'preserveHistoricalCycles' | 'claimsContraceptiveEffectiveness'> = {
  preserveHistoricalCycles: true,
  claimsContraceptiveEffectiveness: false,
};

export function lifeStageLabel(mode: LifeStageMode): string {
  return LABELS[mode];
}

export function capabilitiesForLifeStage(mode: LifeStageMode): LifeStageCapabilities {
  switch (mode) {
    case 'pregnancy':
      return {
        ...BASE,
        predictNextPeriod: false,
        showFertilityObservations: false,
        showPregnancyLogging: true,
      };
    case 'tryingToConceive':
      return {
        ...BASE,
        predictNextPeriod: true,
        showFertilityObservations: true,
        showPregnancyLogging: false,
      };
    case 'postpartum':
    case 'breastfeeding':
    case 'menopauseTransition':
      return {
        ...BASE,
        predictNextPeriod: false,
        showFertilityObservations: false,
        showPregnancyLogging: false,
      };
    case 'cycleTracking':
    case 'perimenopause':
    case 'hormonalContraception':
      return {
        ...BASE,
        predictNextPeriod: true,
        showFertilityObservations: false,
        showPregnancyLogging: false,
      };
  }
}
